Implement Job Management Package

rchammonds opened the pull request:
First PR for the challenge!

Implements a "job" package which handles the job management portion of the system. It exposes the API defined in the design doc, with the exception of `job.Stdout/Stderr` methods which return `(io.ReaderCloser, error)` instead of just a `io.ReadCloser` as those methods are fallible. 

You'll also find a `streamer` package, which provides a utility for reading from a file that has an active writer (a "live" file). This streamer implements the `io.ReadCloser` interface and automatically returns an io.EOF when the writer closes the file. I hand cranked my own inotify utility to implement this, since `fsnotify` could not return an event on file close.

As you've probably guessed, the `streamer` package handles most of the concerns around streaming stdout/stderr. The ability to detect a file close notification is extremely useful, as it means I don't have to propagate any sort of "done" notification from the job to the streamer. The `streamer.LiveFileStreamer` implementation is decoupled from the job package and stands on its own. 

Run tests with `make test` and/or `make test-race`.

rchammonds committed:
First push of jobs package plus helper packages for handling concurrent reads to a 'live' file. Happy path tests for job behavior are working.

rchammonds committed:
Bug fixes, finally fix EOL settings, a few more tests, and the beginnings of a Makefile

rchammonds committed:
Refactor the watcher a bit for testability (and readability too). Add some white box tests for watcher (both happy-path and a couple error cases)

rchammonds committed:
Clean up some error handling when creating output readers. Add some more comments too.

rchammonds committed:
update readme

zmb3 commented on Makefile:
```diff
@@ -0,0 +1,12 @@
+	@go test -v ./...
+
+# requires c compiler
+test-race: echo
```
`test` and `test-race` should also be `PHONY` targets, right?

zmb3 commented on testdata/testprograms/echo:
Mistakenly committed?

zmb3 commented on job/job.go:
```diff
@@ -0,0 +1,230 @@
+type JobArgs struct {
+	Command    string
+	Args       []string
+	StdoutPath string
+	StderrPath string
```
Should these be exported? From my perspective, the fact that you're using files to power the output streaming functionality is an implementation detail and not something you want consumers of the package to rely on.

rchammonds replied:
They should not. Looks like I got a little fast and loose with exported/unexported fields.

Actually, same with the `streamer.LiveFileStreamer`.  The `File` and `WriteWatcher` fields should probably be private since the implementation takes ownership of those objects after calling the constructor.

rchammonds replied:
I misunderstood earlier.  I'm intentionally asking the caller to specify these paths because they might be interested in controlling where the process output ends up. 

For example, my gRPC server will use this package with the expectation that no job data will persist beyond a restart of the server, let alone a reboot of the machine, so it directs output to `/tmp`. Maybe another consumer wants that ensure that the output data survives a reboot. I think the flexibility could be useful.

zmb3 replied:
Ok, we can leave it as-is for the purpose of this project, but I do think it's leaking implementation details.

The core function that the library is meant to expose is the ability to stream output. Ideally, you could completely change the implementation under the hood (say from writing to files to writing to an in-memory buffer) without the client knowing. That's hard to do once your public API exposes the fact that we're writing to files.

rchammonds replied:
I think you've swung me. I'm coming at it from the perspective that a caller might want to direct stdout/stderr to different locations depending on the job... But if that's what the caller wants to do, they can implement it themselves by calling `job.Stdout` and directing the stream wherever they please. 

We could also extend the job API to include `job.ClearOutput` method to clean up any memory/files related to the job. That way the caller can free up any resources related to output management (files and/or memory) without knowing the implementation details.

zmb3 commented on job/job.go:
```diff
@@ -0,0 +1,230 @@
+}
+
+type Job struct {
+	sync.Mutex
```
I don't love embedding the mutex like this, because you make the `Lock` and `Unlock` functions part of the job's public API. Another package that consumes your job library needs to only call `Lock` at the wrong time to get your library to deadlock.

rchammonds replied:
Great point. I'll refactor the mutex to an unexported field.

zmb3 commented on job/job.go:
```diff
@@ -0,0 +1,230 @@
+	if errors.Join(err, err2) != nil {
+		logFileClose(stdoutFile)
+		logFileClose(stderrFile)
+		return nil, fmt.Errorf("error creating output file(s): %w", err)
```
Do you intentionally only wrap `err` and not `err2` here?

rchammonds replied:
No 😩 . I should be grabbing the result of `errors.Join` and wrapping that.

zmb3 commented on job/job_test.go:
```diff
@@ -0,0 +1,152 @@
+
+	bldr := strings.Builder{}
+	for idx := range count {
+		bldr.WriteString(prefix)
+		bldr.WriteString(" ")
+		bldr.WriteString(strconv.Itoa(idx + 1))
+		bldr.WriteString("\n")
```
Alternatively:
```suggestion
		fmt.Fprintln(bldr, "%v %d", prefix, idx+1)
```

zmb3 commented on job/job.go:
```diff
@@ -0,0 +1,230 @@
+	}
+}
+
+func NewJob(args JobArgs) (*Job, error) {
```
`job.NewJob` stutters a bit. How about `job.New` instead?

zmb3 commented on streamer/streamer.go:
No changes requested, but this might make sense as an `internal/` package to avoid leaking implementation details outside the library.

zmb3 commented on streamer/streamer.go:
```diff
@@ -0,0 +1,114 @@
+func (l *LiveFileStreamer) fileRead(p []byte) (int, error) {
+	count, err := l.File.Read(p)
+	if err != nil {
+		// Read returns 0, io.EOF, so we need not check count
+		// or deal with data in 'p' on EOF
```
Are you sure? Godocs say this:

```
When Read encounters an error or end-of-file condition after successfully reading n > 0 bytes, it returns the number of bytes read. It may return the (non-nil) error from the same call or return the error (and n == 0) from a subsequent call. An instance of this general case is that a Reader returning a non-zero number of bytes at the end of the input stream may return either err == EOF or err == nil. The next Read should return 0, EOF.

Callers should always process the n > 0 bytes returned before considering the error err. Doing so correctly handles I/O errors that happen after reading some bytes and also both of the allowed EOF behaviors.
```

rchammonds replied:
That's true for the more generic contract defined for an `io.Reader`. In this instance, we're reading from an `os.File` though, where the documentation for the `Read` method says:
`At end of file, Read returns 0, io.EOF.`

So the `os` package has more strict semantics for file reads.

rchammonds replied:
Though, I guess this could be a footgun for a future developer that changes this `os.File` to an `io.Reader` without realizing the differences.

zmb3 replied:
> So the os package has more strict semantics for file reads.

Thanks for pointing this out. I'm fine with leaving this as-is. It's a short project, so we're less concerned about footguns for future developers than we might otherwise be.

zmb3 commented on streamer/streamer.go:
```diff
@@ -0,0 +1,114 @@
+
+func (l *LiveFileStreamer) Read(p []byte) (int, error) {
+	if l.done != nil {
+		return 0, l.done
```
Doesn't the `io.Reader` contract require that we return `io.EOF` when done?

rchammonds replied:
It does for the happy path, but implementations need some way to communicate read errors. I believe it's acceptable to return non-EOF errors when an implementation encounters an internal error.

zmb3 replied:
Right, but you want to try to stick to the semantics of `io.Reader` as much as possible so that Go's io just works.

For example, the docs for `io.Copy` say:

>A successful Copy returns err == nil, not err == EOF. Because Copy is defined to
> read from src until EOF, it does not treat an EOF from Read as an error to be reported.

If your streamer doesn't return `io.EOF` when it reaches the end of the stream, then `io.Copy` will return an error instead of succeeding, which can have consequences as an unexpected error bubbles up the stack.

One way to induce this behavior is to mimic a job that never writes any output. We can create a `LiveFileStreamer` pointed at an empty file, close it, and then try to `io.Copy` from it. The expected behavior would be that 0 bytes are copied and no error is returned, but with the current implementation we'll get an error.

rchammonds replied:
I've pushed some whitebox tests for the `streamer` package demonstrating that `LiveFileStreamer` is (to the best of my knowledge) meeting the `io.Reader` contract. One of the test uses `io.Copy` to copy data from the streamer into a buffered writer and shows that subsequent calls return `0, io.EOF`.

I'm also using other `io` utilities like `ReadAll` and `ReadFull` in a handful of tests for both the streamer and job package.
Am I still missing an edge case? I'll take a closer look after work this afternoon, but I'm having trouble spotting anything.

zmb3 commented on streamer/streamer.go:
```diff
@@ -0,0 +1,114 @@
+		return 0, l.done
+	}
+
+	if !l.useWatch {
+		return l.fileRead(p)
+	}
```
I see what you're doing now, and it probably works, but I find it a bit hard to follow since the core logic is split across `fileRead` and `Read` and `fileRead` is leaving behind state to tell the next read how to behave.

I wonder if we can express the same logic in a simpler way where it's all in a single function and not split up like this.

rchammonds replied:
I'm happy to take a stab at refactoring it! Agreed that the logic is a bit unwieldy, and this function went through several iterations as I migrated from `fsnotify` to my own `FileWatcher`. Let me give it another shot.

rchammonds replied:
Cleaned this up quite a bit with my last commit. Let me know what you think.

zmb3 replied:
Looks much better, thanks for the update.

rosstimothy commented on Makefile:
```diff
@@ -0,0 +1,12 @@
+
+# requires c compiler
+test-race: echo
+	CGO_ENABLED=1 go test -race ./...
```
Is CGO required for the race detector? Is there any reason to have two separate test targets?

rchammonds replied:
Unfortunately, yes. And that's why I broke it up into two separate targets. If you're on a system that doesn't a C compiler ready to go you can at least still run the tests without the race detector.

rosstimothy commented on job/job.go:
```diff
@@ -0,0 +1,230 @@
+				// the SIGKILL, but it was *probably* us
+				// Maybe the documentation around the "STOPPED" state should instead be
+				// that this state means the program was forcefully terminated, rather than
+				// terminated *by a user*
```
If this transition was handled in a different layer, or was enriched with information from the gRPC service you could probably infer better that a process was stopped manually by a human.

rosstimothy commented on job/job.go:
```diff
@@ -0,0 +1,230 @@
+func logFileClose(f *os.File) {
+	if err := f.Close(); err != nil {
+		if !errors.Is(err, os.ErrInvalid) {
+			slog.Error(fmt.Sprintf("Failed to close file: %s", err))
```
Suggestion: embrace structured logging 
```suggestion
			slog.Error("Failed to close file", "error", err)
```

rosstimothy commented on job/job.go:
```diff
@@ -0,0 +1,230 @@
+	var exited bool
+	j.Lock()
+	exited = j.processExited
+	exitErr = j.exitErr
```
Since this is a pointer are there any concerns with interacting with it after releasing the lock?

rchammonds replied:
Ah.. yes there is and I should have just made a copy of `exitErr` to return to the caller. I'll get this fixed.

rosstimothy commented on job/job.go:
```diff
@@ -0,0 +1,230 @@
+	var err error
+	j.Lock()
+	if !j.processExited {
+		err = j.cmd.Process.Kill()
```
Should the lock be held while issuing the kill signal? Perhaps rewriting this to exit early if the process already exited would clean up some of the error handling here as well?

rchammonds replied:
I think we definitely want to hold the lock while doing this especially considering the server implementation may receive multiple concurrent requests to stop the job. The documentation doesn't explicitly say that `Process.Kill` is safe for concurrent calls, which usually means it is not. I also don't see any warnings in the documentation about this function potentially blocking which leads me to believe that we won't be holding up anyone else for too terribly long.

rosstimothy commented on job/job.go:
```diff
@@ -0,0 +1,230 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to create watcher: %w", err)
+	}
+
+	closeWatcher := false
+	j.Lock()
+	if j.processExited {
+		// Process has exited, which means the watcher *may* not see
+		// the 'close' event that it needs to clean itself up. So we'll close it ourselves.
+		// After all 'processExited == true' implies that the final 'write' to the output files
+		// have completed
+		closeWatcher = true
+	}
+	// else processExited == false
+	// we can't release the lock and *then* create the watcher - the output
+	// files may get closed *just before* we initilaize the watcher.
+	// That's why we create the watcher ahead of time above ^
+	j.Unlock()
+
+	readHandle, err := os.Open(path)
+	if err != nil {
+		// need to clean up the watcher since we've failed to initialize
+		// a LiveFileStreamer
+		closeWatcher = true
+		err = fmt.Errorf("error opening output file '%s' for reading: %w", path, err)
+	}
+
+	if closeWatcher {
+		_ = watcher.Close() // close is safe to be called multiple times
+		for range watcher.Events() {
+		} // Drain watcher events
+	}
+
+	if err != nil {
+		// err has already been wrapped with contextual message
+		return nil, err
+	}
+
+	return streamer.NewLiveFileStreamer(readHandle, watcher), nil
+}
```
Are there any changes that could be made to the streamer package to make the API easier to use for callers? There is a lot of sublty and nuance going on in this function to accomodate the streamer package.

rchammonds replied:
I had to think on it for a few minutes, but yea, I have an idea.

I could write a constructor `streamer.NewLiveFileStreamerFromPath(path string)` that handles creation of the read handle and watcher internally. I'll also add a `StopWatching` method to close the watcher owned by the `LiveFileStreamer`. 

The order of operations becomes pretty simple:
- Create the LiveFileStreamer
- Lock the job and inspect exit status
- if exited, stop the watcher, otherwise do nothing (we don't need to hold the lock while calling `StopWatching``)

The internal `Watcher` is completely hidden from the caller, which is probably how it should've been from the start. If that sounds alright with you, I'll implement this when I get some time this afternoon.

rosstimothy replied:
I like making the watcher an internal implementation detail, though I might suggest going one step further. Can the job provide the streamer package with the path _and_ wether the job is still running? Then everything related to the watcher can remain internal.

rchammonds replied:
It's easy enough to give the LiveFileStreamer/Watcher a channel to listen on to determine if the process (and therefore the writer) has exited. I think I just got tunnel vision on having the streamer/watcher stop itself when the writer closes the file. 

Listening for the job to simply close a channel is easy enough. I didn't need to roll my own watcher implementation on top of inotify then, but that's alright. It was an interesting learning experience.

rchammonds replied:
```
func (j *Job) watchOutput(path string) (io.ReadCloser, error) {
	fileStreamer, err := streamer.NewLiveFileStreamer(path, j.processDone)
	if err != nil {
		return nil, fmt.Errorf("failed to create file streamer: %w", err)
	}
	return fileStreamer, nil
}
```

That's quite a bit simpler

rosstimothy replied:
Much simpler 👍!

rchammonds committed:
Address some of the low hanging fruit of PR feedback
- Mark .PHONY targets as such
- unexport internal 'newState' function
- Track successful attempts to stop a job with the 'userKilled' bool
- unembed job lock
- "embrace structured logging"
- Fix unused result of errors.Join
- Fix read of 'ExitError' after releasing lock
- Rename 'NewJob' to 'New'

rchammonds committed:
Implement the more complicate feedback from the PR.
- Simplify LiveFileStreamer API and provide an explicit cancellation signal from the job upon process exit
- Change semantics of 'FileWriteWatcher'. It now listens for file write events until explicity closed by the caller
- Fix incorrect shutdown sequence of 'FileWriteWatcher'. Remove the watch first, then let the reader goroutine clean up/close the watch file descriptor
- Simplify the 'Read' method of 'LiveFileStreamer'. The read method no longer needs to track state via the streamer instance

rchammonds committed:
Less verbose write to string builder

Co-authored-by: Zac Bergquist <zbergquist99@gmail.com>

rchammonds committed:
'Fprintf' instead of 'Fprintln'

zmb3 commented on streamer/streamer_test.go:
```diff
@@ -0,0 +1,149 @@
+func NewTestResources(dir string) (TestResources, error) {
+	var allErrs error
+	testFile, err := os.OpenFile(filepath.Join(dir, "test"), os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0666)
+	allErrs = errors.Join(allErrs, err)
```
What's going on with the `errors.Join` here. Won't you have at most 1 error here on any invocation?

rchammonds replied:
Yes. Previously I was initializing a couple other resources in this utility and so the join had purpose. I mistakenly left this over after refactoring last night. I'll get this cleaned up.

rosstimothy commented on job/job.go:
```diff
@@ -0,0 +1,202 @@
+		return JobstatusComplete
+	}
+	return JobStatusRunning
+}
```
Can this be simplified down to the following?
```suggestion
	if !processExited {
		return JobStatusRunning
	}
	
	if userKilled {
		return JobStatusStopped
	}
	
	return JobstatusComplete
}
```

rchammonds replied:
That would probably be fine. I had assumed (maybe incorrectly) that there was a possibility that the process could still exit gracefully after the call to `process.Kill`. And for that reason, I think there might be value in doing our due diligence to inspect the process state and determine if it was indeed stopped via a kill signal or had a graceful exit. 

It's a very tight window, and likely won't matter much to the user. I'm open to simplifying to this suggestion.

rchammonds replied:
Implemented

rosstimothy commented on job/job.go:
```diff
@@ -0,0 +1,202 @@
+		if !errors.Is(err, os.ErrInvalid) {
+			slog.Error("Failed to close file", "error", err)
+		}
+	}
```
If `ErrInvalid` is only returned when the file is nil perhaps return early in that case and drop the special error handling?
```suggestion
	if f == nil {
		return
	}
	
	if err := f.Close(); err != nil {
		slog.Error("Failed to close file", "error", err)
	}
```

rosstimothy commented on job/job.go:
```diff
@@ -0,0 +1,202 @@
+
+		close(newJob.processDone)
+		newJob.processExited = true
+		newJob.exitErr, _ = err.(*exec.ExitError)
```
Suggestion: use errors.As?

rosstimothy commented on job/job.go:
```diff
@@ -0,0 +1,202 @@
+func createOutputFile(path string) (*os.File, error) {
+	// We need to open a file for writing, create if not exists,
+	// and truncate existing files
+	flags := os.O_CREATE | os.O_WRONLY | os.O_TRUNC
```
```suggestion
	const flags = os.O_CREATE | os.O_WRONLY | os.O_TRUNC
```

rchammonds committed:
Add whitebox tests to streamer to demonstrate that LiveFileStreamer is meeting the io.Reader contract in both happy-path and error cases

rchammonds committed:
better error validation in new streamer tests

rchammonds committed:
simplify job state calculation per PR feedback

rchammonds committed:
simplify logging around file close errors

rchammonds committed:
clean up 'ExitErr' usage. Set with 'errors.As'. Add const qualifier to file flags.

nklaassen commented on streamer/streamer.go:
```diff
@@ -0,0 +1,102 @@
+	// Indicates that the file will receive no more writes
+	writerDone chan struct{}
+
+	// manage close behavior
+	closeOnce *sync.Once
```
nit: i don't think this needs to be a pointer, if it's a plain value of type sync.Once there's no need to initialize it.

nklaassen commented on streamer/streamer_whitebox_test.go:
```diff
@@ -0,0 +1,74 @@
+	writeHandle, err := os.CreateTemp(t.TempDir(), "")
+	require.NoError(t, err)
+	defer writeHandle.Close()
+	//os.OpenFile(filepath.Join(t.TempDir(),
```
```suggestion
```

nklaassen commented on streamer/streamer_whitebox_test.go:
```diff
@@ -0,0 +1,74 @@
+	// Reading should fail with a non-EOF error
+	_, err = io.ReadAll(fs)
+	require.Error(t, err)
+	require.False(t, errors.Is(err, io.EOF))
```
what if i told you `require.NotErrorIs(t, err, io.EOF)` exists

rchammonds replied:
I would start using it, that's what!

rchammonds committed:
Last PR feedback
